flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/table/UpsertPulsarTableITCase.java

```diff
@@ -0,0 +1,574 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.pulsar.table;
+
+import org.apache.flink.table.api.TableResult;
+import org.apache.flink.table.planner.factories.TestValuesTableFactory;
+import org.apache.flink.test.junit5.MiniClusterExtension;
+import org.apache.flink.types.Row;
+
+import org.junit.Assert;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+
+import java.time.Duration;
+import java.time.LocalDateTime;
+import java.util.*;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.connector.pulsar.table.testutils.PulsarTableTestUtils.collectRows;
+import static org.apache.flink.connector.pulsar.table.testutils.PulsarTableTestUtils.comparedWithKeyAndOrder;
+import static org.apache.flink.connector.pulsar.table.testutils.PulsarTableTestUtils.waitingExpectedResults;
+import static org.apache.flink.table.planner.factories.TestValuesTableFactory.changelogRow;
+import static org.apache.flink.table.utils.TableTestMatchers.deepEqualTo;
+
+/**
+ * IT cases for the Pulsar table source and sink. It aims to verify runtime behaviour and certain
+ * use cases are correct and can produce/consume the desired records as user specifies.
+ */
+@ExtendWith(MiniClusterExtension.class)
+public class UpsertPulsarTableITCase extends PulsarTableITCase {
+    private static final String USERS_TOPIC = "users";
+    private static final String WORD_COUNT_TOPIC = "word_count";
+
+    @ParameterizedTest
+    @ValueSource(strings = {JSON_FORMAT})
+    public void testAggregate(String format) throws Exception {
+        String topic = WORD_COUNT_TOPIC + "_" + format;
+        pulsar.operator().createTopic(topic, 4);
+        // -------------   test   ---------------
+        wordCountToUpsertPulsar(format, topic);
+        wordFreqToUpsertPulsar(format, topic);
+    }
+
+    /**
+     * There is a known issue when parallelism is set to 2 in the temporal join. This is related to
+     * the watermark of Pulsar source. This requires further investigation.
+     */
+    @ParameterizedTest
+    @ValueSource(strings = {JSON_FORMAT})
+    public void testTemporalJoin(String format) throws Exception {
+        String topic = USERS_TOPIC + "_" + format;
+        pulsar.operator().createTopic(topic, 2);
+
+        env.setParallelism(1);
+        writeChangelogToUpsertPulsarWithMetadata(format, topic);
+        env.setParallelism(1);
```

Comment:
Here the parallelism should be set to 2, but the test would fail if set to 2. My guess is the split assignment caused one source parallelism to be idle and couldn't trigger the join.